### PR TITLE
Added action_rules field to ActionMessage; refactored dataclasses int…

### PR DESCRIPTION
**Pull request by eriksendc**
Summary
---------

This PR moves the generations of classification labels to the end of the pdq matcher lambda. In addition, a refactor is included to organize dataclasses in smaller, targeted modules (root reason to do this was cyclical dependencies).

This PR introduces a list of action rules into the action message class so that the rule(s) that is the source of the action being taken is eventually able to be persisted once the action is taken by the action performer. This work can be reviewed in the second commit.

While adding the action rules field to the action message, I found myself writing more .to_aws and .from_aws code. Instead of continuing down that road, I pivoted and tried out HasAWSSerialization. It worked with a couple of minor additions. This work can be reviewed in the first commit.

Other PRs for context:
* [PR 532 - Actions add action rule to action message](https://github.com/facebook/ThreatExchange/pull/532)
* [PR 528 - ActionRule: list to set refactor for must and must not have labels](https://github.com/facebook/ThreatExchange/pull/528)
* [PR 519 - Implemented action evaluation along with some tests](https://github.com/facebook/ThreatExchange/pull/519)
* [PR 512 - Messages for actions and reactions queues](https://github.com/facebook/ThreatExchange/pull/512)

Test Plan
---------

0. python -m py.test
1. `terraform destroy`
2. `make upload_docker`
3. `terraform apply`
4. `python -m hmalib.scripts.populate_config_db load_example_configs`
5. upload "mini-castle.jpg"
6. wait for hashing, matching and actioning to occur
7. validate that the EnqueueMiniCastleForReview WebhookPostActionPerformer is exercised (https://webhook.site/#!/01cef721-bdcc-4681-8430-679c75659867) and that the payload delivered has the complete action message:

```json
{
  "content_key": "images/mini-castle.jpg",
  "content_hash": "361da9e6cf1b72f5cea0344e5bb6e70939f4c70328ace762529cac704297354a",
  "matching_banked_signals": [
    {
      "banked_content_id": "4169895076385542",
      "bank_id": "303636684709969",
      "bank_source": "te",
      "classifications": [
        {
          "key": "BankedContentIDClassification",
          "value": "4169895076385542"
        },
        {
          "key": "Classification",
          "value": "true_positive"
        },
        {
          "key": "BankIDClassification",
          "value": "303636684709969"
        },
        {
          "key": "BankSourceClassification",
          "value": "te"
        }
      ]
    }
  ],
  "action_label": {
    "key": "Action",
    "value": "EnqueueMiniCastleForReview"
  },
  "action_rules": [
    {
      "name": "Enqueue Mini-Castle for Review",
      "action_label": {
        "key": "Action",
        "value": "EnqueueMiniCastleForReview"
      },
      "must_have_labels": [
        {
          "key": "Classification",
          "value": "true_positive"
        },
        {
          "key": "BankIDClassification",
          "value": "303636684709969"
        }
      ],
      "must_not_have_labels": [
        {
          "key": "BankedContentIDClassification",
          "value": "3364504410306721"
        }
      ]
    }
  ]
}
```

### CODE REVIEW

File: hasher-matcher-actioner/hmalib/common/actioner_models.py [hunk 1]
```diff
@@ -5,129 +5,14 @@
...
+from hmalib.common.message_models import BankedSignal
 from hmalib.common.logging import get_logger
-from hmalib.models import BankedSignal, MatchMessage
+from hmalib.common.message_models import MatchMessage
 from requests import get, post, put, delete, Response
 
 logger = get_logger(__name__)
 
 
-@dataclass(unsafe_hash=True)
```
Comment (eriksendc) [inline hasher-matcher-actioner/hmalib/common/actioner_models.py:15]: Moved to classification_models.py

File: hasher-matcher-actioner/hmalib/common/actioner_models.py [hunk 2]
```diff
@@ -5,129 +5,14 @@
...
-    key: str = field(default="BankIDClassification", init=False)
-
-
-@dataclass(unsafe_hash=True)
-class BankedContentIDClassificationLabel(Label):
-    key: str = field(default="BankedContentIDClassification", init=False)
-
-
-@dataclass(unsafe_hash=True)
-class ActionLabel(Label):
```
Comment (eriksendc) [inline hasher-matcher-actioner/hmalib/common/actioner_models.py:47]: Moved to evaluator_labels.py

File: hasher-matcher-actioner/hmalib/common/actioner_models.py [hunk 3]
```diff
@@ -5,129 +5,14 @@
...
-    action_label: ActionLabel
-    must_have_labels: t.Set[Label]
-    must_not_have_labels: t.Set[Label]
-
-
-TUrl = t.Union[t.Text, bytes]
-
-
-@dataclass
-class ActionMessage(MatchMessage):
```
Comment (eriksendc) [inline hasher-matcher-actioner/hmalib/common/actioner_models.py:82]: moved to message_models.py

File: hasher-matcher-actioner/hmalib/lambdas/actions/action_evaluator.py [hunk 4]
```diff
@@ -133,27 +139,6 @@ def get_actions_to_take(
     return action_label_to_action_rules
 
 
-def get_classifications_by_match(match_message: MatchMessage) -> t.List[t.Set[Label]]:
```
Comment (eriksendc) [inline hasher-matcher-actioner/hmalib/lambdas/actions/action_evaluator.py:136]: Logically speaking this moves to pdq_matcher.py

File: hasher-matcher-actioner/hmalib/lambdas/pdq/pdq_matcher.py [hunk 5]
```diff
@@ -124,7 +121,7 @@ def lambda_handler(event, context):
                         str(signal_id), str(privacy_group), str(metadata["source"])
                     )
                     for tag in metadata["tags"].get(privacy_group, []):
-                        banked_signal.classifications.append(tag)
+                        banked_signal.add_classification(tag)
```
Comment (eriksendc) [inline hasher-matcher-actioner/hmalib/lambdas/pdq/pdq_matcher.py:124]: this call to `add_classification()` does the work that was removed from action_evaluator.py.

File: hasher-matcher-actioner/hmalib/models.py [hunk 6]
```diff
@@ -338,40 +337,3 @@ def from_time_range(
             & Key("UpdatedAt").between(start_time, end_time),
             ProjectionExpression=cls.DEFAULT_PROJ_EXP,
         ).get("Items", [])
-
-
-@dataclass
-class BankedSignal:
```
Comment (eriksendc) [inline hasher-matcher-actioner/hmalib/models.py:344]: moved to message_models.py

**Review by schatten — APPROVED** [review]
From the looks of it, and as you mentioned, most of the diff is just moving things around and that looks good to me.

**Pull request merged**